### ISSUE TITLE
Denoising SDXL iteration images for coherent image previews that a user could understand

### ISSUE THREAD

**Issue by MonkeeMan1** [huggingface/diffusers#7001]
Hello,

I'm currently trying to create image previews with SDXL. This works! However, the image output are very noisy. A very long time ago I found a solution to this for sd1.5 but unfortunately it has been lost to time. 

How would I go about denoising these images so they are a little more coherent to a human viewer? I know the first couple of iterations are always going to be very noisy, but eventually it should be possible to convert this noise into a blurry image that a human could understand. 

```py
import time
from diffusers import StableDiffusionXLPipeline
import torch
import matplotlib.pyplot as plt

pipe = StableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
).to("cuda")

```python
prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"

def callback(pipe, step_index, timestep, callback_kwargs):
    latents = callback_kwargs.get("latents")

    start_time = time.time()
    with torch.no_grad():
        pipe.upcast_vae()
        latents = latents.to(
            next(iter(pipe.vae.post_quant_conv.parameters())).dtype)
        images = pipe.vae.decode(
            latents / pipe.vae.config.scaling_factor, return_dict=False)[0]
        images = pipe.image_processor.postprocess(images, output_type='pil')

        images[0].save(f"./imgs/{step_index}.png")

    end_time = time.time()

    print(f"Time taken to generate image: {end_time - start_time} seconds")

    return callback_kwargs

pipe(prompt=prompt, callback_on_step_end=callback)
```

**Comment by yiyixuxu** [comment]
is this what you're looking for? https://github.com/huggingface/diffusers/discussions/6991#discussioncomment-8491149
for questions like this, can we use discussion in the future? https://github.com/huggingface/diffusers/discussions

**Comment by MonkeeMan1** [comment]
Hey, thank you very much for the reply. I apologise that this is the wrong place to put this question, in the future I will definetly make it in the discussions section. 

Unfortunately this isn't quite what I'm looking for. The image previews with this solution are still very noisy. A previous solution I had with sd 1.5 looked like the image attached below:

![image](https://github.com/huggingface/diffusers/assets/98274841/85750133-f577-4709-863c-3c18d5f4c2aa)


This would be ideal as it is really amazing to see how the images improve with no noise in them.

**Comment by CoffeeVampir3** [comment]
Heya, I think I have an example of what you're asking about here <https://gist.github.com/CoffeeVampir3/610e4627042ac8f36b45da6ec3af776f>

basically there's one extra step where you decode the latents at each step, this can be kind of slow so this example uses TAESDXL vae decoder <https://github.com/madebyollin/taesd>

**Comment by MonkeeMan1** [comment]
> Heya, @MonkeeMan1 I think I have an example of what you're asking about here https://gist.github.com/CoffeeVampir3/610e4627042ac8f36b45da6ec3af776f This notebook is a bit old so may not run, but should serve as an example of how to do the thing.
> 
> Basically there's one extra step where you decode the latents at each step, this can be kind of slow so this example uses TAESDXL vae decoder https://github.com/madebyollin/taesd

Hey, thank you very much for the response. Taesdxl definetly looks like this is what im looking for. However, the implementation you sent doesn't quite seem to do it for me. I may just be making a mistake, so apologies if thats the case. I've took out the relevant stuff (I think) and this should work as far as I understand it. 

The output from this code can be seen below;
```python
import io
from diffusers import DiffusionPipeline, LMSDiscreteScheduler, AutoencoderTiny
import numpy as np
import torch
from PIL import Image


pipe = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16,
).to("cuda")

scheduler = LMSDiscreteScheduler(
    beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", num_train_timesteps=1000)


TINY_AUTOENCODER = AutoencoderTiny.from_pretrained(
    "madebyollin/taesd", torch_dtype=torch.float16)
TINY_AUTOENCODER.to("cuda")


prompt = "A capybara holding a sword whilst wearing a knights costuem,"


def to_png_image(img_np):
    """Convert a numpy array to PNG format image."""
    img = Image.fromarray((img_np * 255).astype(np.uint8))
    buf = io.BytesIO()
    img.save(buf, format='png', compress_level=0)
    return buf.getvalue()


def decode_tensors(pipe, step, timestep, callback_kwargs):
    latents = callback_kwargs["latents"]
    img = TINY_AUTOENCODER.decode(latents)
    img_np = img[0].squeeze(0).permute(
        1, 2, 0).cpu().detach().numpy().astype('float32')
    img_np = np.clip((img_np + 1) / 2.0, 0, 1)
    buf = to_png_image(img_np)
    with open(f"./imgs/{step}.png", 'wb') as f:
        f.write(buf)

    return callback_kwargs


image = pipe(
    height=1024,
    width=1024,
    prompt=prompt,
    negative_prompt="",
    guidance_scale=7.5,
    num_inference_steps=20,
    callback_on_step_end=decode_tensors,
    callback_on_step_end_tensor_inputs=["latents"],
).images[0]

image.save("./imgs/final.png")
```
Output:
![image](https://github.com/huggingface/diffusers/assets/98274841/e976a7f3-c877-4095-bd0d-627285813b90)

desired output should resemble a blury image something like this:
![image](https://github.com/huggingface/diffusers/assets/98274841/6dd2f6c2-bbde-450c-a969-5270b6d06a73)

**Comment by MonkeeMan1** [comment]
Hi, I am still looking for a solution to this problem if anybody could help :)

**Comment by jbbqqf** [comment]
Hi! Triaging older issues this week — looks like this one fits better as a discussion than a tracked bug.

- The issue already carries the `should-move-to-discussion` label set by the maintainers — see the [issue](https://github.com/huggingface/diffusers/issues/7001).
- @yiyixuxu earlier linked an existing discussion thread and asked that follow-ups happen on [Discussions](https://github.com/huggingface/diffusers/discussions).
- @CoffeeVampir3 shared a [working gist](https://gist.github.com/CoffeeVampir3/610e4627042ac8f36b45da6ec3af776f) that demonstrates the requested image-preview pattern (decode latents with TAESDXL each step).

Would you mind closing this and continuing on Discussions if needed? Happy to be wrong if there's a concrete diffusers bug here.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually.)
